### ISSUE TITLE
Create a test framework

### ISSUE THREAD

**Issue by SanderMertens** [cortoproject/corto#45]
In order to formalize testing and increase overall productivity when writing tests, a supportive framework would be nice. Of course there is a plethora of testing frameworks out there, so one could ask why build one from scratch?

For a number of reasons. 
 1. A test framework based on Cortex would work for any programming language that has a binding with the platform. As such, it would ensure conformity across languages in how testcases are build, organized and log their output.

 2. The upcoming web bindings and documentation features would complement the test framework  perfectly. Basically, the test framework would be extended with realtime visual feedback for free.

 3. In future releases bindings will be added that allow building truly distributed applications with Cortex. This in essence would enable the framework - again for free - to act as a distributed test engine.

 4. In future releases functionality will be added to the platform that allows for persistent storage of objects. This would, for free, add the capability of the framework to keep track of historic test results.

 5. Keeping the test results around in the Cortex object store will allow users to write scripts based on this data.

In summary, this framework will eventually be language agnostic, scriptable, provide real-time visual feedback on the execution of testcases, can manage running testcases on multiple machines simultaneously and store historical results. 

That is combination of features that I have not yet seen with any OTS framework. The framework in itself would become a good reason to use the platform.

The framework would not be used just for testing the platform, but should also be used for testing applications build by end users.

**Comment by jleeothon** [comment]
The current use of setUp and tearDown methods is an ugly hack. Any inherited setUp and tearDown are executed before the overridden one, instead of properly overriding it, making some really weird behavior.

Instead, it could be proposed to "annotate" setUp and tearDown methods as a procedure extending method. Like this, it is a more sensible that any parent setUp and tearDown methods are discovered and run, instead of discovered and run but not overridden.

**Comment by jleeothon** [comment]
The current test runner suffers from the issue that individual test cases' executions are not isolated. So, if test case A changes object ::foo and test case B also changes object ::foo; one would have to be careful, e.g. with the test order (baaaad) or cleaning it up in tearDown (unDRY).

Some requirements worth nothing:

- The test runner is able to isolate test case executions.
- We want an easy command line instruction to run tests (e.g. `rake test`).

We would want a different object store for each test. This means that a different object store is loaded and unloaded for each test.
- Is this easy to do from C?, i.e.: load a store, run a test, get and accumulate the results, unload the store. Sounds hard.
- An alternative could be running different cortex commands, that is, the rakefile will look something like
```
output = `cortex mytests.cx`
success = if $? == 0 then true else false end
```
- However, the above approach will keep every child process completely separate, and the results couldn't easily be collected in a same object store.
    - Talk about keeping the results; the current TestRunner only collects results numerically; it doesn't hold the executed test cases--they're anonymous, not under its scope--, we could change that, but it's not relevant now.
    - Also, how can we tell a cortex object store which specific test to run from the command line?

**Comment by SanderMertens** [comment]
Test framework is implemented. Additional features for the test framework will be captured by new issues.
